File: snake/snake.zig
const std = @import("std");

const Direction = enum(u2) { UP = 0, DOWN = 1, RIGHT = 2, LEFT = 3 };

const Snake = struct { pos_x: u16, pos_y: u16, s: ?*Snake };
const Food = struct { pos_x: u16, pos_y: u16 };

var prng: std.Random.Xoshiro256 = undefined;
var win: std.c.winsize = undefined;
var snakeDirection: Direction = Direction.RIGHT;

pub fn main() !void {
    var input: [5]u8 = undefined;
    try updatewinsize();

    // Initialize random numbers.
    prng = std.Random.DefaultPrng.init(blk: {
        var seed: u32 = undefined;
        try std.posix.getrandom(std.mem.asBytes(&seed));
        break :blk seed;
    });

    // Initialize allocator.
    var AreaAlloc: std.heap.ArenaAllocator = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer AreaAlloc.deinit();
    var allocator: std.mem.Allocator = AreaAlloc.allocator();

    // Allocate snake
    const player: *Snake = try allocator.create(Snake);
    player.* = Snake{ .pos_x = 1, .pos_y = 1, .s = null };

    // Allocate food.
    var food: []Food = try allocator.alloc(Food, 10);
    f_init(&food);

    // Initialize termios.
    const defaultTerm: std.posix.termios = try std.posix.tcgetattr(std.posix.STDIN_FILENO);
    defer _ = std.posix.tcsetattr(std.posix.STDIN_FILENO, std.posix.TCSA.NOW, defaultTerm) catch {};
    var newTerm: std.posix.termios = defaultTerm;
    newTerm.lflag.ECHO = false;
    newTerm.lflag.ICANON = false;
    try std.posix.tcsetattr(std.posix.STDIN_FILENO, std.posix.TCSA.NOW, newTerm);

    // Main game loop.
    while (true) {
        try clearwin();
        try updatewinsize();
        // TODO: Read only if there is input (Epoll).

        input = undefined;
        _ = try std.posix.read(std.posix.STDIN_FILENO, &input);

        if (std.mem.count(u8, &input, "\x1b[A") > 0 or std.mem.count(u8, &input, "w") > 0) {
            snakeDirection = Direction.UP;
        } else if (std.mem.count(u8, &input, "\x1b[B") > 0 or std.mem.count(u8, &input, "s") > 0) {
            snakeDirection = Direction.DOWN;
        } else if (std.mem.count(u8, &input, "\x1b[C") > 0 or std.mem.count(u8, &input, "d") > 0) {
            snakeDirection = Direction.RIGHT;
        } else if (std.mem.count(u8, &input, "\x1b[D") > 0 or std.mem.count(u8, &input, "a") > 0) {
            snakeDirection = Direction.LEFT;
        }

        try s_move(player);
        if (try checkCollisions(player, &food, allocator)) {
            break;
        }

        try f_print(&food);
        try s_print(player);

        if (std.mem.count(u8, &input, "q") > 0 or std.mem.count(u8, &input, "\x1b") > 0) {
            break;
        }
        std.posix.nanosleep(0, 4 * 33333333);
    }
    return;
}

// FUNCTIONS //

/// Clears the terminal.
inline fn clearwin() !void {
    _ = try std.posix.write(std.posix.STDOUT_FILENO, "\x1b[2J");
    return;
}

/// Updates the global 'win' struct to have the latest window size.
inline fn updatewinsize() !void {
    const out: c_int = std.c.ioctl(std.posix.STDOUT_FILENO, std.posix.T.IOCGWINSZ, &win);
    if (out != 0) {
        const err = std.posix.errno(out);
        std.debug.print("C errno: {}\n", .{err});
        return error.FailedCFunctionCall;
    }
    return;
}

/// Moves the cursor to column: x, row: y on the terminal.
inline fn movecursor(x: u16, y: u16) !void {
    var tmp: [10]u8 = undefined;
    _ = try std.fmt.bufPrint(&tmp, "\u{001b}[{d};{d}H", .{ y, x });
    _ = try std.posix.write(std.posix.STDOUT_FILENO, &tmp);
    return;
}

/// Check collisions.
fn checkCollisions(s: *Snake, f: *[]Food, SnakeAllocator: std.mem.Allocator) !bool {
    var AreaAlloc: std.heap.ArenaAllocator = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer AreaAlloc.deinit();
    var allocator: std.mem.Allocator = AreaAlloc.allocator();

    const occupied = try allocator.alloc([]bool, win.ws_row);
    for (occupied, 0..) |_, i| {
        occupied[i] = try allocator.alloc(bool, win.ws_col);
    }

    // Get occupied cells of the snake.
    var curr: ?*Snake = s.s;
    while (curr) |snake| {
        if (snake.pos_y > 0 and snake.pos_y < win.ws_row and snake.pos_x > 0 and snake.pos_x < win.ws_col) {
            occupied[snake.pos_y][snake.pos_x] = true;
        }
        curr = snake.s;
    }

    // Get occupied cells by the food.
    for (f.*) |*food| {
        var random = prng.random();
        if (food.pos_x > 0 and food.pos_x < win.ws_col and food.pos_y > 0 and food.pos_y < win.ws_row) {
            occupied[food.pos_y][food.pos_x] = true;
        } else {
            // Get out of bound food within boundaries
            food.pos_x = random.int(u16) % win.ws_col;
            food.pos_y = random.int(u16) % win.ws_row;
            occupied[food.pos_y][food.pos_x] = true;
        }
    }

    // Check if there was collision at all.
    if (s.pos_x > 0 and s.pos_x < win.ws_col and s.pos_y > 0 and s.pos_y < win.ws_row and occupied[s.pos_y][s.pos_x]) {
        // Check food collision.
        for (f.*) |*food| {
            var random = prng.random();
            if (s.pos_x == food.pos_x and s.pos_y == food.pos_y) {
                food.pos_x = random.int(u16) % win.ws_col;
                food.pos_y = random.int(u16) % win.ws_row;
                try s_add(s, SnakeAllocator);
                return false;
            }
        }

        // Check snake-to-snake collision.
        curr = s.s;
        while (curr) |snake| {
            if (s.pos_x == snake.pos_x and s.pos_y == snake.pos_y) {
                try movecursor(0, 0);
                _ = try std.posix.write(std.posix.STDOUT_FILENO, " ! Death ! ");
                return true; // EXIT
            }
            curr = snake.s;
        }
    }

    // Boundry check.
    curr = s;
    while (curr) |snake| {
        if (snake.pos_x < 0) {
            snake.pos_x = win.ws_col;
        } else if (snake.pos_x > win.ws_col) {
            snake.pos_x = 0;
        }
        if (snake.pos_y < 0) {
            snake.pos_y = win.ws_row;
        } else if (snake.pos_y > win.ws_row) {
            snake.pos_y = 0;
        }
        curr = snake.s;
    }

    return false;
}

/// Initialize the given food array.
fn f_init(f: *[]Food) void {
    for (f.*) |*food| {
        var random = prng.random();
        food.pos_x = random.int(u16) % win.ws_col;
        food.pos_y = random.int(u16) % win.ws_row;
    }
    return;
}

/// Print all food items.
pub fn f_print(f: *[]Food) !void {
    for (f.*) |*food| {
        try movecursor(food.pos_x, food.pos_y);
        _ = try std.posix.write(std.posix.STDOUT_FILENO, "\u{001b}[48;5;2m*\u{001b}[0m\n");
    }

    return;
}

/// Add a segment to the snake.
fn s_add(s: *Snake, alloc: std.mem.Allocator) !void {
    var curr: ?*Snake = s.s;
    while (curr) |snake| {
        curr = snake.s;
    }

    curr = s;
    if (curr) |snake| {
        const old_x: u16 = snake.pos_x;
        const old_y: u16 = snake.pos_y;
        try s_move(s);

        snake.s = try alloc.create(Snake);
        snake.s.?.* = Snake{ .pos_x = old_x, .pos_y = old_y, .s = null };
    }
    return;
}

/// Move the snake.
fn s_move(s: *Snake) !void {
    var curr: ?*Snake = s;
    var x: u16 = undefined;
    var y: u16 = undefined;

    if (curr) |segment| {
        x = segment.pos_x;
        y = segment.pos_y;
    }

    switch (snakeDirection) {
        Direction.UP => {
            y -= 1;
        },
        Direction.DOWN => {
            y += 1;
        },
        Direction.RIGHT => {
            x += 1;
        },
        Direction.LEFT => {
            x -= 1;
        },
    }

    // https://betterexplained.com/articles/swap-two-variables-using-xor/
    while (curr) |segment| {
        segment.pos_x = segment.pos_x ^ x;
        x = segment.pos_x ^ x;
        segment.pos_x = segment.pos_x ^ x;
        segment.pos_y = segment.pos_y ^ y;
        y = segment.pos_y ^ y;
        segment.pos_y = segment.pos_y ^ y;
        curr = segment.s;
    }
    return;
}

/// Print the snake with all it's segments.
fn s_print(s: *Snake) !void {
    var curr: ?*Snake = s;
    while (curr) |snake| {
        try movecursor(snake.pos_x, snake.pos_y);
        _ = try std.posix.write(std.posix.STDOUT_FILENO, "\u{001b}[48;5;7mS\u{001b}[0m");
        curr = snake.s;
    }
    return;
}

File: snake/snake.c
// clang snake.c -std=gnu99 -O3 -g -Wextra -Wall

#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

// My snake is indeed a linked list.
struct snake {
  int pos_x;
  int pos_y;
  struct snake *s;
};

// Food struct for points ect.
struct food {
  int pos_x;
  int pos_y;
};

enum direction { UP = 0, DOWN = 1, RIGHT = 2, LEFT = 3 };

void winclear(void);
void movecursor(int x, int y);
void nsleep(int nsec);
void updatewinsize(void);

int check_collisions(struct snake *s, struct food *f, int size);

// Dunctions for the food.
void f_print(struct food *f, int size);
void f_init(struct food *f, int size);

// Functions for the snake.
void s_move(struct snake *s);
void s_add(struct snake *s);
void s_delete(struct snake *s);
void s_print(struct snake *s);

// Used to get terminal size.
struct winsize win;
int snake_dir = RIGHT;

int main(void) {
  srand(time(NULL));
  char input[5];

  // Get initial terminal size.
  updatewinsize();

  // Initialize the food struct.
  struct food food[10];
  int food_size = sizeof(food) / sizeof(struct food);
  f_init(food, food_size);

  // Initialize the player struct.
  struct snake *player = malloc(sizeof(struct snake));
  (*player) = (struct snake){0, 0, NULL};
  s_add(player);

  // Disable echo and canonical mode.
  struct termios default_t, new_t;
  tcgetattr(STDIN_FILENO, &default_t);
  new_t = default_t;
  new_t.c_lflag &= ~(ECHO | ICANON);
  tcsetattr(STDIN_FILENO, TCSANOW, &new_t);

  // Set the standard input to non-blocking mode
  int flags = fcntl(STDIN_FILENO, F_GETFL, 0);
  fcntl(STDIN_FILENO, F_SETFL, flags | O_NONBLOCK);

  while (1) {
    winclear();
    updatewinsize();

    // Get input.
    memset(&input, '\0', sizeof(input));      // 'clear' the input buffer.
    read(STDIN_FILENO, input, sizeof(input)); // Get input (POSIX).

    if (strcmp(input, "\x1b[A") == 0) {
      snake_dir = UP;
    } else if (strcmp(input, "\x1b[B") == 0) {
      snake_dir = DOWN;
    } else if (strcmp(input, "\x1b[C") == 0) {
      snake_dir = RIGHT;
    } else if (strcmp(input, "\x1b[D") == 0) {
      snake_dir = LEFT;
    }

    s_move(player);
    if (check_collisions(player, food, food_size) == 1) {
      s_delete(player);
      break;
    }

    f_print(food, food_size);
    s_print(player);

    nsleep(2 * 33333333); // Should be read as: x / 30fps

    if (strcmp(input, "\x1b") == 0) {
      s_delete(player);
      break;
    }
  }

  tcsetattr(STDIN_FILENO, TCSANOW, &default_t);
  return EXIT_SUCCESS;
}

//////////////// FUNCTION DEFINITIONS ////////////////

void winclear(void) {
  write(STDOUT_FILENO, "\x1b[2J", 4);
  return;
}

void movecursor(int x, int y) {
  char buf[12];
  snprintf(buf, sizeof(buf), "\x1b[%d;%dH", y, x);
  write(STDOUT_FILENO, buf, strlen(buf));
  return;
}

void nsleep(int nsec) {
  struct timespec tm = {0, nsec};
  nanosleep(&tm, NULL);
  return;
}

void updatewinsize(void) {
  ioctl(STDIN_FILENO, TIOCGWINSZ, &win);
  return;
}

int check_collisions(struct snake *s, struct food *f, int size) {
  bool **occupied = calloc(win.ws_row, sizeof(bool *));
  for (int i = 0; i < win.ws_row; i++) {
    occupied[i] = calloc(win.ws_col, sizeof(bool));
  }

  // Get occupied cells of the snake.
  struct snake *curr = s->s;
  while (curr != NULL) {
    if (curr->pos_y > 0 && curr->pos_y < win.ws_row && curr->pos_x > 0 &&
        curr->pos_x < win.ws_col) {
      occupied[curr->pos_y][curr->pos_x] = true;
    }
    curr = curr->s;
  }

  // Get occupied cells by the food.
  for (int i = 0; i < size; i++) {
    if (f[i].pos_x > 0 && f[i].pos_x < win.ws_col && f[i].pos_y > 0 &&
        f[i].pos_y < win.ws_row) {
      occupied[f[i].pos_y][f[i].pos_x] = true;
    } else {
      // Get out of bound food within boundaries
      f[i].pos_x = rand() % win.ws_col;
      f[i].pos_y = rand() % win.ws_row;
      occupied[f[i].pos_y][f[i].pos_x] = true;
    }
  }

  // Check if there was collision at all.
  if (s->pos_x > 0 && s->pos_x < win.ws_col && s->pos_y > 0 &&
      s->pos_y < win.ws_row && occupied[s->pos_y][s->pos_x]) {
    // Check food collision.
    for (int i = 0; i < size; i++) {
      if (s->pos_x == f[i].pos_x && s->pos_y == f[i].pos_y) {
        f[i].pos_x = rand() % win.ws_col;
        f[i].pos_y = rand() % win.ws_row;
        s_add(s);
        for (int i = 0; i < win.ws_row; i++) {
          free(occupied[i]);
        }
        free(occupied);
        return 0;
      }
    }

    // Check snake-to-snake collision.
    curr = s->s;
    while (curr != NULL) {
      if (s->pos_x == curr->pos_x && s->pos_y == curr->pos_y) {
        movecursor(0, 0);
        write(STDOUT_FILENO, " ! Death ! ", 11);
        for (int i = 0; i < win.ws_row; i++) {
          free(occupied[i]);
        }
        free(occupied);
        return 1;
      }
      curr = curr->s;
    }
  }

  // Boundry check.
  curr = s;
  while (curr != NULL) {
    if (curr->pos_x < 0) {
      curr->pos_x = win.ws_col;
    } else if (curr->pos_x > win.ws_col) {
      curr->pos_x = 0;
    }
    if (curr->pos_y < 0) {
      curr->pos_y = win.ws_row;
    } else if (curr->pos_y > win.ws_row) {
      curr->pos_y = 0;
    }
    curr = curr->s;
  }

  for (int i = 0; i < win.ws_row; i++) {
    free(occupied[i]);
  }
  free(occupied);
  return 0;
}

void f_print(struct food *f, int size) {
  size_t bufsize = size * (10 + 14);
  // 10 for movecursor() 14 for formated output string.
  char *outbuf = malloc(bufsize);
  int index = 0;

  // movecursor(f[i].pos_x, f[i].pos_y);
  for (int i = 0; i < size; i++) {
    index +=
        snprintf(outbuf + index, bufsize - index,
                 "\x1b[%d;%dH\x1b[48;5;2m*\x1b[0m\n", f[i].pos_y, f[i].pos_x);
  }
  write(STDOUT_FILENO, outbuf, strlen(outbuf));

  free(outbuf);
  return;
}

void f_init(struct food *f, int size) {
  for (int i = 0; i < size; i++) {
    f[i].pos_x = rand() % win.ws_col;
    f[i].pos_y = rand() % win.ws_row;
  }
  return;
}

void s_move(struct snake *s) {
  struct snake *curr = s;
  int x = curr->pos_x;
  int y = curr->pos_y;

  switch (snake_dir) {
  case 0:
    y--;
    break;
  case 1:
    y++;
    break;
  case 2:
    x++;
    break;
  case 3:
    x--;
    break;
  }

  // https://betterexplained.com/articles/swap-two-variables-using-xor/
  while (curr != NULL) {
    curr->pos_x = curr->pos_x ^ x;
    x = curr->pos_x ^ x;
    curr->pos_x = curr->pos_x ^ x;
    curr->pos_y = curr->pos_y ^ y;
    y = curr->pos_y ^ y;
    curr->pos_y = curr->pos_y ^ y;
    curr = curr->s;
  }
  return;
}

void s_add(struct snake *s) {
  struct snake *curr = s;
  while (curr->s != NULL) {
    curr = curr->s;
  }

  int old_x = curr->pos_x;
  int old_y = curr->pos_y;

  s_move(s);

  curr->s = malloc(sizeof(struct snake));
  *curr->s = (struct snake){old_x, old_y, NULL};

  return;
}

void s_delete(struct snake *s) {
  struct snake *curr = s;
  struct snake *nxt;

  while (curr != NULL) {
    nxt = curr->s;
    free(curr);
    curr = nxt;
  }
  return;
}

void s_print(struct snake *s) {
  struct snake *curr = s;

  while (curr != NULL) {
    movecursor(curr->pos_x, curr->pos_y);
    write(STDOUT_FILENO, "\x1b[48;5;7mS\x1b[0m", 14);
    curr = curr->s;
  }
  return;
}
